pandas/_libs/src/inference.pyx

```diff
@@ -606,9 +611,16 @@ cpdef bint is_bool_array(ndarray values):
         if n == 0:
             return False
 
-        for i in range(n):
-            if not util.is_bool_object(objbuf[i]):
-                return False
+        if skipna:
+            for i in range(n):
+                val = objbuf[i]
+                if not util._checknull(val) and not util.is_bool_object(val):
+                    return False
```

Comment:
> Refactoring the entire set of functions is kind of a can of worms, because the fundamental problem here is that we must assume that arrays contain potentially mixed types or objects, which are really the same thing.

Yeah, that's fair.  I was imagining something like this:
~~~python
cpdef check_non_na(ndarray[object] values, <callable> checker):
   n = len(values)
   for i in range(n):
       val = values[i]
       if not util._checknull(val) and not checker(val):
              return False
   return True
~~~